src/Common/AssemblyLoadingUtils.cs

```diff
@@ -0,0 +1,249 @@
+                    continue;
+
+                env.ComponentCatalog.RegisterAssembly(a);
+            }
+        }
+
+        private static string CreateTempDirectory()
+        {
+            string dir = GetTempPath();
+            Directory.CreateDirectory(dir);
+            return dir;
+        }
+
+        private static string GetTempPath()
+        {
+            Guid guid = Guid.NewGuid();
+            return Path.GetFullPath(Path.Combine(Path.GetTempPath(), "MLNET_" + guid.ToString()));
+        }
+
+        private static readonly string[] _filePrefixesToAvoid = new string[] {
+            "api-ms-win",
+            "clr",
+            "coreclr",
+            "dbgshim",
+            "ext-ms-win",
+            "microsoft.bond.",
+            "microsoft.cosmos.",
+            "microsoft.csharp",
+            "microsoft.data.",
+            "microsoft.hpc.",
+            "microsoft.live.",
+            "microsoft.platformbuilder.",
+            "microsoft.visualbasic",
+            "microsoft.visualstudio.",
+            "microsoft.win32",
+            "microsoft.windowsapicodepack.",
+            "microsoft.windowsazure.",
+            "mscor",
+            "msvc",
+            "petzold.",
+            "roslyn.",
+            "sho",
+            "sni",
+            "sqm",
+            "system.",
+            "zlib",
+        };
+
+        private static bool ShouldSkipPath(string path)
+        {
+            string name = Path.GetFileName(path).ToLowerInvariant();
+            switch (name)
+            {
+                case "cqo.dll":
+                case "fasttreenative.dll":
+                case "libiomp5md.dll":
+                case "libvw.dll":
+                case "matrixinterf.dll":
+                case "microsoft.ml.neuralnetworks.gpucuda.dll":
+                case "mklimports.dll":
+                case "microsoft.research.controls.decisiontrees.dll":
+                case "microsoft.ml.neuralnetworks.sse.dll":
+                case "neuraltreeevaluator.dll":
+                case "optimizationbuilderdotnet.dll":
+                case "parallelcommunicator.dll":
+                case "microsoft.ml.runtime.runtests.dll":
+                case "scopecompiler.dll":
+                case "tbb.dll":
+                case "internallearnscope.dll":
+                case "unmanagedlib.dll":
+                case "vcclient.dll":
+                case "libxgboost.dll":
+                case "zedgraph.dll":
+                case "__scopecodegen__.dll":
+                case "cosmosClientApi.dll":
+                    return true;
+            }
+
+            foreach (var s in _filePrefixesToAvoid)
+            {
+                if (name.StartsWith(s))
+                    return true;
+            }
+
+            return false;
+        }
+
+        private static void LoadAssembliesInDir(IHostEnvironment env, string dir, bool filter)
+        {
+            if (!Directory.Exists(dir))
+                return;
```

Comment:
Why do you silently return? Why not throw or at least debug assert.

Reply:
Because there are cases where this is called on potentially non-existing directories, for example [here](https://github.com/dotnet/machinelearning/pull/970/files/430c4a78d1c2a07780715126c3c123a338f160e6#diff-cac9e327be3e6627783bc09ab3f1c749R227) where we try calling it on the `AutoLoad` folder, which may not exist.  (again that was all existing behavior that I'm refactoring out of ComponentCatalog.)